kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -42,47 +47,157 @@
+    private Long lastOffset;
+    private Integer shallowCount;
+    private Boolean offsetsMonotonic;
+    private Boolean hasProducerId;
+    private Boolean isTransaction;
+    private Long lastOffsetOfFirstBatch;
+    private Integer validBytes;
+}
 
 /**
  * An append-only log for storing messages. Mapping to Kafka Log.scala.
  */
 @Slf4j
 @AllArgsConstructor
 public class PartitionLog {
+
     private final KafkaServiceConfiguration kafkaConfig;
     private final Time time;
     private final TopicPartition topicPartition;
     private final String namespacePrefix;
     private final String fullPartitionName;
     private final EntryFormatter entryFormatter;
-    private final Optional<TransactionCoordinator> transactionCoordinator;
+    private final ProducerStateManager producerStateManager;
+
+    /**
+     * AppendOrigin is used mark the data origin.
+     */
+    public enum AppendOrigin {
+        Coordinator,
+        Client,
+        Log
+    }
+
+    public AnalyzeResult analyzeAndValidateProducerState(MemoryRecords records,
+                                                         Optional<Long> firstOffset,
+                                                         AppendOrigin origin) {
+        Map<Long, ProducerAppendInfo> updatedProducers = Maps.newHashMap();
+        List<CompletedTxn> completedTxns = Lists.newArrayList();
+
+        for (RecordBatch batch : records.batches()) {
+            if (batch.hasProducerId()) {
+                if (origin.equals(AppendOrigin.Client)) {
+                    Optional<ProducerStateEntry> maybeLastEntry =
+                            producerStateManager.lastEntry(batch.producerId());
+
+                    // if this is a client produce request,
+                    // there will be up to 5 batches which could have been duplicated.
+                    // If we find a duplicate, we return the metadata of the appended batch to the client.
+                    if (maybeLastEntry.isPresent()) {
+                        Optional<BatchMetadata> maybeDuplicate =
+                                maybeLastEntry.get().findDuplicateBatch(batch);
+                        if (maybeDuplicate.isPresent()) {
+                            return new AnalyzeResult(updatedProducers, completedTxns, maybeDuplicate);
+                        }
+                    }
+                }
+                // We cache offset metadata for the start of each transaction. This allows us to
+                // compute the last stable offset without relying on additional index lookups.
+                Optional<CompletedTxn> maybeCompletedTxn =
+                        updateProducers(batch, updatedProducers, firstOffset, origin);
+                maybeCompletedTxn.ifPresent(completedTxns::add);
+            }
+        }
+        return new AnalyzeResult(updatedProducers, completedTxns, Optional.empty());
+    }
+
+    public void append(AnalyzeResult analyzeResult, long startOffset, LogAppendInfo appendInfo) {
```

Comment:
It should be a private method. In addition, this method name is confused. It should be a callback, but `append` is similar to `appendRecords`. I recommend the name `publishComplete` and move the other logic into this method. 

Actually, I think there is no need to add a new method. I tried to add the code into `offsetFuture.whenComplete` directly like

```java
                final long lastOffset = offset + numMessages - 1;
                analyzeResult.updatedProducers().forEach((pid, producerAppendInfo) -> {
                    if (log.isDebugEnabled()) {
                        log.debug("Append pid: [{}], appendInfo: [{}], lastOffset: [{}]",
                                pid, producerAppendInfo, lastOffset);
                    }
                    // When we have real start offset, update current txn first offset.
                    producerAppendInfo.updateCurrentTxnFirstOffset(appendInfo.isTransaction(), offset);
                    producerStateManager.update(producerAppendInfo);
                });
                analyzeResult.completedTxns().forEach(completedTxn -> {
                    // update to real last offset
                    completedTxn.lastOffset(lastOffset);
                    long lastStableOffset = producerStateManager.lastStableOffset(completedTxn);
                    producerStateManager.updateTxnIndex(completedTxn, lastStableOffset);
                    producerStateManager.completeTxn(completedTxn);
                });
```

Reply:
Addressed.